src/rbac/defaultRoleManager.ts

```diff
@@ -179,20 +181,34 @@ export class DefaultRoleManager implements RoleManager {
     this.domainMatchingFunc = fn;
   }
 
+  /**
+   * addDomainHierarchy sets a rolemanager to define role inheritance between domains
+   * @param rm RoleManager to define domain hierarchy
+   */
+  public async addDomainHierarchy(rm: RoleManager): Promise<void> {
+    this.hasDomainHierarchy = true;
+    this.domainHierarchyManager = rm;
+  }
+
   private generateTempRoles(domain: string): Roles {
     loadOrDefault(this.allDomains, domain, new Roles());
 
-    const patternDomain = new Set([domain]);
-    if (this.hasDomainPattern) {
+    const extraDomain = new Set([domain]);
+    if (this.hasDomainPattern || this.hasDomainHierarchy) {
       this.allDomains.forEach((value, key) => {
-        if (this.domainMatchingFunc(domain, key)) {
-          patternDomain.add(key);
+        if (this.hasDomainPattern && this.domainMatchingFunc(domain, key)) {
+          extraDomain.add(key);
+        } else if (this.hasDomainHierarchy) {
+          if (!this.domainHierarchyManager.syncedHasLink) throw Error('domain hierarchy must be syncronous');
```

Comment:
This is a little bit problematic part of the current enforcer implementation. Currently there is no way to know whether this should be run in a promise or not. Maybe we should set a isAsync parameter when `hasLink` is called and turn it false before `hasLink` returns.

Reply:
At the moment I am unsure whether it even matters if defaultRoleManager's hasLink is async or not. More about it here: https://github.com/casbin/node-casbin/pull/378/files#r944271038